### DISCUSSION TITLE
Any examples on how to work with @psalm-type from within a plugin?

### DISCUSSION THREAD

**Discussion by rzvc** [vimeo/psalm#7977] [Plugin development]
Like the title says, I'm looking for some examples to get me going.

The whole documentation around the plugin section could use some love. It's very difficult to piece things together currently.

**Comment by orklah** [comment]
Yeah, documentation is pretty sparse... Just looking at the code, it seems like '@psalm-types' are stored in FileStorage::$type_aliases.

Not sure what you want to do, but if you have a file path, you can retrieve type aliases with `$codebase->file_storage_provider->get($file_path)`
If you have a class storage, you should be able to retrieve its path through `ClassLikeStorage::$location`

Can you tell me more about what you want to achieve if you need more info?

**Reply by rzvc** [reply]
Hi,

Thanks for the answer!

I'm trying to check whether a string is a valid type in the context of the calling function/method.

It's not exactly clear to me what is the scope of `@psalm-types` and what its availability is. For example, in here it's available to all functions, but not to the other classes: https://psalm.dev/r/bb71879460

I guess my followup question would be, how do I know which types are available to me, in a given context (from a plugin's perspective)?

**Reply by orklah** [reply]
Your scope example is weird indeed. I don't think type aliases should ever be allowed for functions like that.

There is no complete list of every available type (mainly because the types are discovered as the analysis progress. Each new analysed line could make you discover new types). However, for a given file `FileStorage::$type_aliases` will give you allowed aliases

**Reply by rzvc** [reply]
Looking them up in the `FileStorage` worked, thank you.

The problem is that now I can't compare them with `UnionTypeComparator::isContainedBy`. I actually managed to expand them into something useful, but only first level aliases, so if it's a complex object, it can't be compared. Any ideas on how to obtain a real type, that can be used for comparison?

Edit: Nevermind, found `TypeParser`.

Thanks again.
